### PR TITLE
kv/RocksDBStore: always release column family handles

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
tested at https://pulpito.ceph.com/kchai-2020-06-28_05:13:09-rados-wip-kefu2-testing-2020-06-28-1028-distro-basic-smithi/5185572/

File: src/kv/RocksDBStore.cc [hunk 1]
```diff
@@ -3051,74 +3038,65 @@ int RocksDBStore::reshard(const std::string& new_sharding, const RocksDBStore::r
...
-    goto cleanup;
+    return -1002;
   }
   env->CreateDir(sharding_def_dir);
-  status = rocksdb::WriteStringToFile(env, new_sharding,
-				      sharding_def_file, true);
-  if (!status.ok()) {
+  if (auto status = rocksdb::WriteStringToFile(env, new_sharding,
+					       sharding_def_file, true);
+      !status.ok()) {
```
Comment (aclamk) [inline src/kv/RocksDBStore.cc:3096]: Since `status` is local, maybe just `!rocksdb::Write...(...).ok()` ?

File: src/kv/RocksDBStore.h [hunk 2]
```diff
@@ -478,11 +478,13 @@ class RocksDBStore : public KeyValueDB {
 private:
   WholeSpaceIterator get_default_cf_iterator();
 
+  using cf_deleter_t = std::function<void(rocksdb::ColumnFamilyHandle*)>;
+  using columns_t = std::map<std::string,
+			     std::unique_ptr<rocksdb::ColumnFamilyHandle,
+					     cf_deleter_t>>;
```
Comment (aclamk) [inline src/kv/RocksDBStore.h:484]: Love that you merged names & handles together.

File: src/kv/RocksDBStore.cc [hunk 3]
```diff
@@ -3051,74 +3038,65 @@ int RocksDBStore::reshard(const std::string& new_sharding, const RocksDBStore::r
...
       }
     }
-    flush_batch();
-  out:
-    delete it;
-    delete bat;
-    return r;
+    if (bat.Count() > 0) {
+      flush_batch(&bat);
+    }
```
Comment (aclamk) [inline src/kv/RocksDBStore.cc:3049]: Nice catch, previously even empty batch was submitted.

File: src/kv/RocksDBStore.cc (hunk 1 above)
Comment (tchaikov) [reply src/kv/RocksDBStore.cc:3096]: @aclamk thanks for your review Adam! i will implement your suggestion in a follow up PR. as some rados suites are failing due to this issue, i am going to merge it as it is.

**Comment by tchaikov** [comment]
jenkins test dashboard backend

**Pull request merged**